src/realm/sync/noinst/client_history_impl.cpp

```diff
@@ -767,6 +768,9 @@ void ClientHistory::update_sync_progress(const SyncProgress& progress, const std
         root.set(s_progress_upload_server_version_iip,
                  RefOrTagged::make_tagged(progress.upload.last_integrated_server_version)); // Throws
     }
+    if (previous_upload_client_version < progress.upload.client_version) {
```

Comment:
During automatic client reset, the Realm being reset steals the client ident and client version of the freshly downloaded Realm and starts from an empty sync history. A client reset operation can completely finish successfully locally, but we don't know if the changes generated by a recovery are validated by the server until the server acknowledges the upload. Since the recovery is done in a single write transaction, the assumption here is that if any upload after a reset succeeds, then the server has accepted the recovery without generating a subsequent reset and there is no cycle.

Reply:
Gotcha. So this is basically the crux of this whole PR. Can you add a comment explaining why this is here? Now that you've explained it, this makes sense but it'd be great to document it in the code since the if statement here is not obviously connected to the action it predicates.